README.md
![banner](https://raw.githubusercontent.com/teixeirazeus/hequest/master/readme_assets/banner.png)
[![Codacy Badge](https://app.codacy.com/project/badge/Grade/ca9452a2e4a745718e59fc8ca54de6e2)](https://www.codacy.com/gh/teixeirazeus/hequest/dashboard?utm_source=github.com\&utm_medium=referral\&utm_content=teixeirazeus/hequest\&utm_campaign=Badge_Grade)[![License](https://raw.githubusercontent.com/teixeirazeus/hequest/3bb5c5adb73020d036563be2a155210562789f22/readme_assets/mit.svg)](https://github.com/teixeirazeus/hequest)

Simple and hackable request for dart.

## Installing

1.  Add dependencies to `pubspec.yaml`

    ```yaml
    dependencies:
        hequest:
            git:
                url: https://github.com/teixeirazeus/hequest
    ```

2.  Run pub get.

    ```shell
    flutter pub get
    ```

3.  Import package.

    ```dart
    import 'package:hequest/hequest.dart';
    ```

## Using

Hequest was created with simplicity in mind to handle multiple API requests.

You start by declaring an object of type `Hequest` and passing the `baseUrl` of your api.

```dart
final githubApi = Hequest(baseUrl: 'https://api.github.com');
```

### GET

```dart
final response = await githubApi.get('/users/teixeirazeus');
```

### POST

```dart
final response = await githubApi.post('/users/teixeirazeus', body: {
    'name': 'Zeus Teixeira',
    });
```

### PUT

```dart
final response = await githubApi.put('/users/teixeirazeus', body: {
    'name': 'Zeus Teixeira',
    });
```

### DELETE

```dart
final response = await githubApi.delete('/users/teixeirazeus');
```

### JWT

All requests support sending jwt token.

To use to use the `WithToken` postfix and pass the token as a parameter.

```dart
final response = await githubApi.getWithToken('/users/teixeirazeus', token);
```

### Tips

It is highly recommended that you use `try catch` to handle errors. Along with a larger abstraction with a class for each endpoint of your api.

```dart
class GithubApi {
  final Hequest _hequest;

  GithubApi(this._hequest);

  Future<Map> getUser(String username) async {
    try {
      final response = await _hequest.get('/users/$username');
      if (response.statusCode == 200) {
        return json.decode(response.body);
      } else if (response.statusCode == 404) {
        throw Exception('User not found');
      } else {
        throw Exception('Failed to load user');
      }
    } catch (e) {
      throw Exception('Error getting user');
    }
  }
}
```

```dart
final hequest = Hequest(baseUrl: 'https://api.github.com');
final githubApi = GithubApi(hequest);

final user = await githubApi.getUser('teixeirazeus');
print(user.toString());
```

## Contributing

Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.


CHANGELOG.md
## 0.0.1

*   initial release

## 0.0.2

*   better documentation

## 0.0.3

*   better documentation

## 0.0.4

*   patch request


lib/hequest.dart
library hequest;

import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:http/http.dart';

class Hequest {
  /// The base URL for the API.
  /// Example: https://api.example.com
  String baseUrl;

  /// Hequest constructor.
  Hequest({
    required this.baseUrl,
  });

  /// GET request
  ///
  /// [route] is the route to the endpoint.
  /// Example: /users
  /// Returns a [Response] object.
  Future<Response> get(String route) async {
    return await http.get(Uri.parse(baseUrl + route));
  }

  /// GET request with token authentication
  ///
  ///
  /// [route] is the route to the endpoint.
  /// Example: /users
  /// [token] is the token to authenticate the request.
  /// Returns a [Response] object.
  Future<Response> getWithToken(String route, String token) async {
    return await http.get(
      Uri.parse(baseUrl + route),
      headers: <String, String>{
        'Authorization': 'Bearer $token',
      },
    );
  }

  /// POST request
  ///
  /// [route] is the route to the endpoint.
  /// Example: /users
  /// [jsonMap] is the JSON map to send in the request body.
  /// Returns a [Response] object.
  Future<Response> post(String route, Map<dynamic, dynamic> jsonMap) async {
    return await http.post(
      Uri.parse(baseUrl + route),
      headers: <String, String>{
        'Content-Type': 'application/json; charset=UTF-8',
      },
      body: jsonEncode(jsonMap),
    );
  }

  /// POST request with token authentication
  ///
  /// [route] is the route to the endpoint.
  /// Example: /users
  /// [token] is the token to authenticate the request.
  /// [jsonMap] is the JSON map to send in the request body.
  /// Returns a [Response] object.
  Future<Response> postWithToken(
      String route, String token, Map<dynamic, dynamic> jsonMap) async {
    return await http.post(
      Uri.parse(baseUrl + route),
      headers: <String, String>{
        'Content-Type': 'application/json; charset=UTF-8',
        'Authorization': 'Bearer $token',
      },
      body: jsonEncode(jsonMap),
    );
  }

  /// PUT request
  ///
  /// [route] is the route to the endpoint.
  /// Example: /users
  /// [jsonMap] is the JSON map to send in the request body.
  /// Returns a [Response] object.
  Future<Response> put(String route, Map<dynamic, dynamic> jsonMap) async {
    return await http.put(
      Uri.parse(baseUrl + route),
      headers: <String, String>{
        'Content-Type': 'application/json; charset=UTF-8'
      },
      body: jsonEncode(jsonMap),
    );
  }

  /// PUT request with token authentication
  ///
  /// [route] is the route to the endpoint.
  /// Example: /users
  /// [token] is the token to authenticate the request.
  /// [jsonMap] is the JSON map to send in the request body.
  /// Returns a [Response] object.
  Future<Response> putWithToken(
      String route, String token, Map<dynamic, dynamic> jsonMap) async {
    return await http.put(
      Uri.parse(baseUrl + route),
      headers: <String, String>{
        'Content-Type': 'application/json; charset=UTF-8',
        'Authorization': 'Bearer $token',
      },
      body: jsonEncode(jsonMap),
    );
  }

  /// PATCH request
  ///
  /// [route] is the route to the endpoint.
  /// Example: /users
  /// [jsonMap] is the JSON map to send in the request body.
  /// Returns a [Response] object.
  Future<Response> patch(String route, Map<dynamic, dynamic> jsonMap) async {
    return await http.patch(
      Uri.parse(baseUrl + route),
      headers: <String, String>{
        'Content-Type': 'application/json; charset=UTF-8'
      },
      body: jsonEncode(jsonMap),
    );
  }

  /// PATCH request with token authentication
  ///
  /// [route] is the route to the endpoint.
  /// Example: /users
  /// [token] is the token to authenticate the request.
  /// [jsonMap] is the JSON map to send in the request body.
  /// Returns a [Response] object.
  Future<Response> patchWithToken(
      String route, String token, Map<dynamic, dynamic> jsonMap) async {
    return await http.patch(
      Uri.parse(baseUrl + route),
      headers: <String, String>{
        'Content-Type': 'application/json; charset=UTF-8',
        'Authorization': 'Bearer $token',
      },
      body: jsonEncode(jsonMap),
    );
  }

  /// DELETE request with token authentication
  ///
  /// [route] is the route to the endpoint.
  /// Example: /users
  /// [token] is the token to authenticate the request.
  /// Returns a [Response] object.
  Future<Response> deleteWithToken(String route, String token) async {
    return await http.delete(
      Uri.parse(baseUrl + route),
      headers: <String, String>{
        'Content-Type': 'application/json; charset=UTF-8',
        'Authorization': 'Bearer $token',
      },
    );
  }

  /// DELETE request
  ///
  /// [route] is the route to the endpoint.
  /// Example: /users
  /// Returns a [Response] object.
  Future<Response> delete(String route) async {
    return await http.delete(Uri.parse(baseUrl + route));
  }

  /// DELETE request with token authentication and body
  ///
  /// [route] is the route to the endpoint.
  /// Example: /users
  /// [token] is the token to authenticate the request.
  /// [jsonMap] is the JSON map to send in the request body.
  /// Returns a [Response] object.
  Future<Response> deleteWithTokenNBody(
      String route, String token, Map<String, dynamic> jsonMap) async {
    return await http.delete(
      Uri.parse(baseUrl + route),
      headers: <String, String>{
        'Content-Type': 'application/json; charset=UTF-8',
        'Authorization': 'Bearer $token',
      },
      body: jsonEncode(jsonMap),
    );
  }
}


example/main.dart
import 'dart:convert';

import 'package:hequest/hequest.dart';

class GithubApi {
  final Hequest _hequest;

  GithubApi(this._hequest);

  Future<Map> getUser(String username) async {
    try {
      final response = await _hequest.get('/users/$username');
      if (response.statusCode == 200) {
        return json.decode(response.body);
      } else if (response.statusCode == 404) {
        throw Exception('User not found');
      } else {
        throw Exception('Failed to load user');
      }
    } catch (e) {
      throw Exception('Error getting user');
    }
  }
}

void main() async {
  final hequest = Hequest(baseUrl: 'https://api.github.com');
  final githubApi = GithubApi(hequest);

  final user = await githubApi.getUser('teixeirazeus');
  // ignore: avoid_print
  print(user.toString());
}
